bottlecap/src/otlp/transform.rs

```diff
@@ -0,0 +1,1144 @@
+
+//     SPAN_TYPE_GENERIC_DB.to_string()
+// }
+
+fn get_otel_attribute_value(attributes: &Vec<KeyValue>, key: &str) -> Option<AnyValue> {
+    for attribute in attributes {
+        if attribute.key == key {
+            return attribute.value.clone();
+        }
+    }
+
+    None
+}
+
+fn otel_value_to_string(value: &any_value::Value) -> String {
+    match value {
+        any_value::Value::StringValue(s) => s.clone(),
+        any_value::Value::BoolValue(b) => b.to_string(),
+        any_value::Value::IntValue(i) => i.to_string(),
+        any_value::Value::DoubleValue(d) => d.to_string(),
+        any_value::Value::BytesValue(b) => hex::encode(b),
+        any_value::Value::ArrayValue(a) => serde_json::to_string(a).unwrap_or_default(),
+        any_value::Value::KvlistValue(kvl) => serde_json::to_string(kvl).unwrap_or_default(),
+    }
+}
+
+/// Returns the matched value as a string from the attributes if found.
+///
+/// The first value found will be returned.
+///
+/// If normalize is true, the value will be normalized.
+#[must_use]
+pub fn get_otel_attribute_value_as_string(
+    attributes: &Vec<KeyValue>,
+    key: &str,
+    normalize: bool,
+) -> String {
+    let mut value = String::new();
+    for attribute in attributes {
+        if attribute.key == key {
+            let Some(any_value) = &attribute.value else {
+                continue;
+            };
+
+            let Some(v) = &any_value.value else {
+                continue;
+            };
+
+            value = otel_value_to_string(v);
+
+            break;
+        }
+    }
+
+    if normalize {
+        normalize_tag(&mut value);
+    }
+
+    value
+}
+
+#[must_use]
+pub fn get_otel_attribute_value_as_string_from_resource_or_span(
+    otel_res: &OtelResource,
+    otel_span: &OtelSpan,
+    key: &str,
+    normalize: bool,
+) -> String {
+    let value = get_otel_attribute_value_as_string(&otel_res.attributes, key, normalize);
+    if !value.is_empty() {
+        return value;
+    }
+
+    get_otel_attribute_value_as_string(&otel_span.attributes, key, normalize)
+}
+
+#[must_use]
+pub fn otel_trace_id_to_dd_id(trace_id: &[u8]) -> (u64, u64) {
+    let lower_order_bits = u64::from_be_bytes(trace_id[8..16].try_into().unwrap_or_default());
+    let higher_order_bits = u64::from_be_bytes(trace_id[0..8].try_into().unwrap_or_default());
+
+    (lower_order_bits, higher_order_bits)
+}
+
+#[must_use]
+pub fn otel_span_id_to_u64(span_id: &[u8]) -> u64 {
+    u64::from_be_bytes(span_id.try_into().unwrap_or_default())
+}
+
+fn get_otel_service(otel_res: &OtelResource, normalize: bool) -> String {
+    let mut service =
+        get_otel_attribute_value_as_string(&otel_res.attributes, KEY_DATADOG_SERVICE, normalize);
+
+    if service.is_empty() {
+        service = String::from("otlpresourcenoservicename");
+    }
+
+    if normalize {
+        normalize_service(&mut service);
+    }
+
+    service
+}
+
+fn get_otel_operation_name(otel_span: &OtelSpan) -> String {
```

Comment:
I hate to ask but – is there a chance that the order of any of these method calls matter? Are we sure this is correct?

Reply:
I'm just following what the Go agent does, same order as there